### PR TITLE
tests: lib: mem_alloc: migrating mem_alloc tests to new ztest API

### PR DESCRIPTION
**Pull request by yerabolu**
Migrating mem_alloc tests to new ztest API.

Part of #47002 effort

Signed-off-by: Spoorthy Priya Yerabolu <spoorthy.priya.yerabolu@intel.com>

### CODE REVIEW

File: tests/lib/mem_alloc/prj_negative_testing.conf [hunk 1]
```diff
@@ -1,3 +1,5 @@
 CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
+CONFIG_NO_OPTIMIZATIONS=y
```
Comment (stephanosio) [inline tests/lib/mem_alloc/prj_negative_testing.conf:3]: There is a big difference between disabling optimisation globally vs. only for some functions of interest.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: tests/lib/mem_alloc/prj_negative_testing.conf (hunk 1 above)
Comment (yerabolu) [reply tests/lib/mem_alloc/prj_negative_testing.conf]: Agreed, was trying a few things while testing and forgot to remove them.

**Comment by enjiamai** [comment]
Hi, @yerabolu , all looks good for me and I verified them in some boards such as k64f, nrf52840, nsim, etc. It works well.

I just have some concern on removing the __no_optimization, I checked why it had added, then saw in this previous commit:  seems like it will affect boards which using the ARC MWDT toolchain. 
```
commit fa41ec052d26e2497fc29e28dc0202317d715f30
Author: Watson Zeng <zhiwei@synopsys.com>
Date:   Mon May 24 16:35:36 2021 +0800

    tests: mem_alloc: workaround aggressive optimization

    As we don't use memory allocated in test_calloc, test_no_mem_malloc,
    and test_no_mem_realloc. malloc call can be optimized away (that really
    happens with ARC MWDT toolchain). That breaks the test. So disable
    optimization for these functions.

    Signed-off-by: Watson Zeng <zhiwei@synopsys.com>
```

I did not have a arcmwdt toolchain to verify it.  Hi @IRISZZW , could you please help to check if this still causes problem now? (if we remove the __no_optimization)

**Comment by stephanosio** [comment]
could set the vars volatile.

**Comment by yerabolu** [comment]
> Hi, @yerabolu , all looks good for me and I verified them in some boards such as k64f, nrf52840, nsim, etc. It works well.
> 
> I just have some concern on removing the __no_optimization, I checked why it had added, then saw in this previous commit: seems like it will affect boards which using the ARC MWDT toolchain.
> 
> ```
> commit fa41ec052d26e2497fc29e28dc0202317d715f30
> Author: Watson Zeng <zhiwei@synopsys.com>
> Date:   Mon May 24 16:35:36 2021 +0800
> 
>     tests: mem_alloc: workaround aggressive optimization
> 
>     As we don't use memory allocated in test_calloc, test_no_mem_malloc,
>     and test_no_mem_realloc. malloc call can be optimized away (that really
>     happens with ARC MWDT toolchain). That breaks the test. So disable
>     optimization for these functions.
> 
>     Signed-off-by: Watson Zeng <zhiwei@synopsys.com>
> ```
> 
> I did not have a arcmwdt toolchain to verify it. Hi @IRISZZW , could you please help to check if this still causes problem now? (if we remove the __no_optimization)

Thank You Enjia for verifying it and looking into it. Yes, i am still looking into the __no_optimization attribute.

**Comment by enjiamai** [comment]
Hi @yerabolu , any update on this PR? Thanks.

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

**Pull request closed without merging**